### DISCUSSION TITLE
[Implemented] Choosing default (qualitative) color palette for plotting functions

### DISCUSSION THREAD

**Discussion by IndrajeetPatil** [Open-Systems-Pharmacology/OSPSuite-R#894] [Ideas]
We need to choose a default color palette for #871. The chosen palette should satisfy the following conditions:

- It should have a sufficient number of colors to handle a significant number of groupings (one color per grouping).
- It should be colorblind-friendly. 

I'd personally recommend either _Paired_ or _Dark2_ palette from `{RColorBrewer}` package.

# Available palettes in RColorBrewer

**Note before** 

This does **not** mean we gain  `{RColorBrewer}` package as an additional hard dependency of `{ospsuite}`. We just need to decide on a palette, and I can port over a vector of colors. No need to rely on that entire package for a single palette.

![image](https://user-images.githubusercontent.com/11330453/165050862-07804ac9-eec8-4213-ae1f-e35d6d2a453b.png)

Tabular summary for these palettes

``` r
RColorBrewer::brewer.pal.info
#>          maxcolors category colorblind
#> BrBG            11      div       TRUE
#> PiYG            11      div       TRUE
#> PRGn            11      div       TRUE
#> PuOr            11      div       TRUE
#> RdBu            11      div       TRUE
#> RdGy            11      div      FALSE
#> RdYlBu          11      div       TRUE
#> RdYlGn          11      div      FALSE
#> Spectral        11      div      FALSE
#> Accent           8     qual      FALSE
#> Dark2            8     qual       TRUE
#> Paired          12     qual       TRUE
#> Pastel1          9     qual      FALSE
#> Pastel2          8     qual      FALSE
#> Set1             9     qual      FALSE
#> Set2             8     qual       TRUE
#> Set3            12     qual      FALSE
#> Blues            9      seq       TRUE
#> BuGn             9      seq       TRUE
#> BuPu             9      seq       TRUE
#> GnBu             9      seq       TRUE
#> Greens           9      seq       TRUE
#> Greys            9      seq       TRUE
#> Oranges          9      seq       TRUE
#> OrRd             9      seq       TRUE
#> PuBu             9      seq       TRUE
#> PuBuGn           9      seq       TRUE
#> PuRd             9      seq       TRUE
#> Purples          9      seq       TRUE
#> RdPu             9      seq       TRUE
#> Reds             9      seq       TRUE
#> YlGn             9      seq       TRUE
#> YlGnBu           9      seq       TRUE
#> YlOrBr           9      seq       TRUE
#> YlOrRd           9      seq       TRUE
```

<sup>Created on 2022-04-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by PavelBal** [comment]
Can we interpolate between the colors of a palette in case we need more colors than the palette provides?

**Reply by IndrajeetPatil** [reply]
> Can we interpolate between the colors of a palette in case we need more colors than the palette provides?

We can't interpolate, but we can maybe choose another palette with more than 12 colors if you expect this to be the case. 
Btw, _tlf_ will recycle colors if the palette runs out, and the expectation is that the users would provide their own palettes in such cases.  

Can there really be more than 12 groupings in realistic workflows? 

There are > 60 qualitative color palettes available in R with > 20 colors each (unclear if or how many of them are colorblind-friendly). 
So it depends on what's the number of colors we want the default palette to have, and a palette can be chosen accordingly.

``` r
library(dplyr)
library(paletteer)

palettes_d_names %>%
  filter(type == "qualitative", length > 20L) %>%
  arrange(desc(length))
#> # A tibble: 62 x 5
#>    package      palette  length type        novelty
#>    <chr>        <chr>     <int> <chr>       <lgl>  
#>  1 palettesForR Named       448 qualitative TRUE   
#>  2 palettesForR Inkscape    431 qualitative TRUE   
#>  3 palettesForR Visibone    336 qualitative TRUE   
#>  4 palettesForR Bears       256 qualitative TRUE   
#>  5 palettesForR Bgold       256 qualitative TRUE   
#>  6 palettesForR Blues       256 qualitative TRUE   
#>  7 palettesForR Borders     256 qualitative TRUE   
#>  8 palettesForR Caramel     256 qualitative TRUE   
#>  9 palettesForR Cascade     256 qualitative TRUE   
#> 10 palettesForR China       256 qualitative TRUE   
#> # ... with 52 more rows
```

<sup>Created on 2022-04-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by PavelBal** [comment]
> We can't interpolate

Actually I do not understand why. A color is just an RGB or a HEX number...

> Can there really be more than 12 groupings in realistic workflows?

Sure. 

> and the expectation is that the users would provide their own palettes in such cases.

No, I would expect it to generate a separate color for no matter how many data sets are provided. Sure, if the user is not happy with the automatic colors, he could choose its own palette. But out of the box, I do not want to think about how many colors I might need.

**Reply by IndrajeetPatil** [reply]
> Actually I do not understand why. A color is just an RGB or a HEX number...

Because interpolating color goes against the fact that it's a qualitative color palette, and not a sequential/diverging one.

![image](https://user-images.githubusercontent.com/11330453/165083713-8a084ce8-7a42-45bc-bbaf-f0920453d977.png)

> No, I would expect it to generate a separate color for no matter how many data sets are provided.

But this is unrealistic. Sure, we can generate an infinite number of colors, but what about shapes (`length(tlf::Shapes)` = 38)? What about linetypes (`length(tlf::Linetypes)` = 7)? We have only a few of those available. 

In your specification, you are mapping dataset to both colors and shapes. You will run out of shapes way before colors. But, honestly, I don't know how could anyone make sense of a visualization with more than 38 datasets plotted.

**Comment by PavelBal** [comment]
We need a way to generate as many distinct colors as required. I don't care if we take some of the palettes (I prefer Set1 or Dark2) and only interpolate/generate new colors if the default palette does not suffice.

Example of a figure with >12 data sets (and therefore more than 12 colors would be required):

https://github.com/Open-Systems-Pharmacology/OSP-Qualification-Reports/blob/master/DDI_Qualification_CYP3A4/images/002_2_Qualification_of_Use_Case_CYP3A4-mediated_DDI/004_plotDDIRatioCMAXResidualsVsObserved.png

**Reply by IndrajeetPatil** [reply]
I am including a palette with 51 colors. I hope nobody is planning to visualize more than 51 datasets in a single plot.

``` r
library(paletteer)
library(scales)

show_col(paletteer_d("ggsci::default_igv"))
```

![](https://i.imgur.com/rFUb7S7.png)

<sup>Created on 2022-04-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>


>Example of a figure with >12 data sets (and therefore more than 12 colors would be required):
>
>https://github.com/Open-Systems-Pharmacology/OSP-Qualification-Reports/blob/master/DDI_Qualification_CYP3A4/images/002_2_Qualification_of_Use_Case_CYP3A4-mediated_DDI/004_plotDDIRatioCMAXResidualsVsObserved.png

🤐
